File: Data Structures & Algorithms/insertionSort/submission-2.cs
// Definition for a pair
// public class Pair {
//     public int Key;
//     public string Value;
//
//     public Pair(int key, string value) {
//         Key = key;
//         Value = value;
//     }
// }
public class Solution {
    public List<List<Pair>> InsertionSort(List<Pair> pairs) {
        var pairsList = new List<List<Pair>>();
        
        if(pairs.Count == 0) return pairsList;

        pairsList.Add(new List<Pair>(pairs));
        for(int i = 1; i < pairs.Count; i++){
            int j = i-1; 
            while(j >= 0 && pairs[j].Key > pairs[j+1].Key){
                var temp = pairs[j+1];
                pairs[j+1] = pairs[j];
                pairs[j] = temp;
                j--;
            }

            pairsList.Add(new List<Pair>(pairs));
        }

        return pairsList;
    }
}

File: Data Structures & Algorithms/mergeSort/submission-2.py
# Definition for a pair.
# class Pair:
#     def __init__(self, key: int, value: str):
#         self.key = key
#         self.value = value
class Solution:
    def mergeSort(self, pairs: List[Pair]) -> List[Pair]:
        return self.MergeSortPairs(pairs, 0, len(pairs) - 1)

    def MergeSortPairs(self, pairs: List[Pair], left: int, right: int) -> List[Pair]:
        if(left < right):
            mid = (left + right) // 2
            self.MergeSortPairs(pairs, left, mid)
            self.MergeSortPairs(pairs, mid + 1, right)
            self.Merge(pairs, left, mid, right)
        return pairs
    
    def Merge(self, pairs: List[Pair], left: int, mid: int, right: int):
        tempRight = []
        tempLeft= []

        for i in range(left, mid + 1):
            tempLeft.append(pairs[i])
        
        for i in range(mid + 1, right + 1):
            tempRight.append(pairs[i])
        
        i = 0
        j = 0
        k = left

        while i < len(tempLeft) and j < len(tempRight):
            if tempLeft[i].key <= tempRight[j].key:
                pairs[k] = tempLeft[i]
                i += 1
            else:
                pairs[k] = tempRight[j]
                j += 1
            k += 1
    
        # Copy remaining elements
        while i < len(tempLeft):
            pairs[k] = tempLeft[i]
            i += 1
            k += 1
    
        while j < len(tempRight):
            pairs[k] = tempRight[j]
            j += 1
            k += 1

        return pairs

        
